### PR TITLE
Add support for specifying process group backend to relevant distributed strategies

### LINKED ISSUE #11725: Support custom process group backends
### Motivation

Find the link there: https://github.com/facebookresearch/fairring

Screenshot 2022-02-03 at 19 21 26

Screenshot 2022-02-03 at 19 20 26

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by ananthsub**
Fixes #11725

### Does your PR introduce any breaking changes? If yes, please list them.
No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -302,6 +302,21 @@ def tpu_distributed() -> bool:
...
+def get_default_process_group_backend_for_device(device: torch.device) -> str:
+    return "nccl" if device.type == "cuda" else "gloo"
+
+
+def _get_process_group_backend_from_env() -> Optional[str]:
+    torch_backend = os.getenv("PL_TORCH_DISTRIBUTED_BACKEND")
+    if torch_backend:
+        rank_zero_deprecation(
+            "Environment variable `PL_TORCH_DISTRIBUTED_BACKEND`"
+            " was deprecated in v1.6 and will be removed in v1.8."
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:314]: What's the argument behind deprecating the environment variable?

File: pytorch_lightning/strategies/fully_sharded.py [hunk 2]
```diff
@@ -107,6 +108,7 @@ def __init__(
             cluster_environment=cluster_environment,
             checkpoint_io=checkpoint_io,
             precision_plugin=precision_plugin,
+            pg_backend=pg_backend,  # TODO: Remove inheritance from DDPStrategy
```
Comment (carmocca) [inline pytorch_lightning/strategies/fully_sharded.py:111]: How does the added TODOs relate to these changes?

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -302,6 +302,21 @@ def tpu_distributed() -> bool:
     return _TPU_AVAILABLE and xm.xrt_world_size() > 1
 
 
+def get_default_process_group_backend_for_device(device: torch.device) -> str:
+    return "nccl" if device.type == "cuda" else "gloo"
+
+
+def _get_process_group_backend_from_env() -> Optional[str]:
+    torch_backend = os.getenv("PL_TORCH_DISTRIBUTED_BACKEND")
+    if torch_backend:
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:311]: Should we be strict here about empty strings?

```suggestion
    if torch_backend is not None:
```

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/distributed.py:314]: The motivation to specify the process group backend was from https://github.com/PyTorchLightning/pytorch-lightning/issues/5355 

The environment variable is not obvious or documented for users. https://github.com/PyTorchLightning/pytorch-lightning/pull/5981

Given it's now explicitly available on the constructor, why would someone use the env var over this argument?

Wouldn't this also be better supported from the LIghtning CLI, given there are other strategy arguments like `find_unused_parameters` that people pass through the constructor too? Why is one option set through a constructor argument and another through an env var?

I advocate deprecating this for consistency

File: pytorch_lightning/strategies/fully_sharded.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/fully_sharded.py:111]: they're not, but are needed for overall API stability. the only reason these arguments extend from DDP is for subprocess launching. I can remove the TODOs and make that a separate issue.

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:314]: I would personally keep both with a preference toward the pg_backend argument, I believe it is quite useful for users to `pip install fairing` and add the flag and they don't need to modify their code.

File: pytorch_lightning/strategies/fully_sharded.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/strategies/fully_sharded.py:111]: Yeah, I would prefer that as there is more nuance to what this TODO says

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/distributed.py:314]: Yes, it does.

The `LightningCLI` supersedes all the argparse logic we had plus the saving part of `save_hyperparameters`, however, not everybody uses or wants to use the CLI so that's why none of that has been deprecated.

File: CHANGELOG.md [hunk 4]
```diff
@@ -391,6 +394,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `pytorch_lightning.utilities.warnings.LightningDeprecationWarning` in favor of `pytorch_lightning.utilities.rank_zero.LightningDeprecationWarning`
 
 
+- Deprecated specifying the process group backend through the environment variable `PL_TORCH_DISTRIBUTED_BACKEND` ([#11745](https://github.com/PyTorchLightning/pytorch-lightning/pull/11745))
+
```
Comment (awaelchli) [inline CHANGELOG.md:398]: we would need to update the docs in gpu.rst as well

File: pytorch_lightning/strategies/parallel.py [hunk 5]
```diff
@@ -136,3 +134,14 @@ def block_backward_sync(self):
     def teardown(self) -> None:
         self.cluster_environment.teardown()
         super().teardown()
+
+    @property
+    def torch_distributed_backend(self) -> str:
```
Comment (awaelchli) [inline pytorch_lightning/strategies/parallel.py:139]: It would be great if we could keep it to the top where all other properties belong

**Review by carmocca — APPROVED** [review]

File: docs/source/accelerators/gpu.rst [hunk 6]
```diff
@@ -264,11 +264,17 @@ Select torch distributed backend
 By default, Lightning will select the ``nccl`` backend over ``gloo`` when running on GPUs.
 Find more information about PyTorch's supported backends `here <https://pytorch.org/docs/stable/distributed.html>`__.
 
-Lightning exposes an environment variable ``PL_TORCH_DISTRIBUTED_BACKEND`` for the user to change the backend.
+Lightning allows explicitly specifying the backend via the `process_group_backend` constructor argument on the relevant Strategy classes.
```
Comment (ananthsub) [inline docs/source/accelerators/gpu.rst:267]:
```suggestion
Lightning allows explicitly specifying the backend via the `process_group_backend` constructor argument on the relevant Strategy classes. By default, Lightning will select the appropriate process group backend based on the hardware used.
```

File: docs/source/accelerators/gpu.rst [hunk 7]
```diff
@@ -264,11 +264,17 @@ Select torch distributed backend
...
-Lightning exposes an environment variable ``PL_TORCH_DISTRIBUTED_BACKEND`` for the user to change the backend.
+Lightning allows explicitly specifying the backend via the `process_group_backend` constructor argument on the relevant Strategy classes. By default, Lightning will select the appropriate process group backend based on the hardware used.
 
-.. code-block:: bash
+.. code-block:: python
+
+    from pytorch_lightning.strategies import DDPStrategy
+
+    # Specify the process group backend
+    ddp = DDPStrategy(process_group_backend="gloo")
```
Comment (ananthsub) [inline docs/source/accelerators/gpu.rst:274]:
```suggestion
    # Explicitly specify the process group backend if you choose to
    ddp = DDPStrategy(process_group_backend="nccl")
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/distributed.py [hunk 8]
```diff
@@ -301,6 +301,21 @@ def tpu_distributed() -> bool:
     return _TPU_AVAILABLE and xm.xrt_world_size() > 1
 
 
+def get_default_process_group_backend_for_device(device: torch.device) -> str:
+    return "nccl" if device.type == "cuda" else "gloo"
+
+
+def _get_process_group_backend_from_env() -> Optional[str]:
```
Comment (four4fish) [inline pytorch_lightning/utilities/distributed.py:308]: why this is a private method?

File: pytorch_lightning/strategies/ddp.py [hunk 9]
```diff
@@ -168,10 +178,15 @@ def setup_distributed(self):
...
-        # set up server using proc 0's ip address
-        # try to init for 20 times at max in case ports are taken
-        # where to store ip_table
-        init_dist_connection(self.cluster_environment, self.torch_distributed_backend)
+        self._process_group_backend = self._get_process_group_backend()
+        init_dist_connection(self.cluster_environment, self._process_group_backend)
+
+    def _get_process_group_backend(self) -> str:
+        return (
+            self._process_group_backend
```
Comment (four4fish) [inline pytorch_lightning/strategies/ddp.py:186]: should we check whether the process_group_backend user passed in is valid or not?

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 9 above)
Comment (ananthsub) [reply pytorch_lightning/strategies/ddp.py:186]: What sort of validation did you have in mind? Currently the validation will happen within init_dist_connection.

File: pytorch_lightning/utilities/distributed.py (hunk 8 above)
Comment (ananthsub) [reply pytorch_lightning/utilities/distributed.py:308]: Since this PR deprecates the env var support, I didn't want to expose a specific public API that would also be deprecated/removed later

File: pytorch_lightning/strategies/ddp.py (hunk 9 above)
Comment (four4fish) [reply pytorch_lightning/strategies/ddp.py:186]: Do you think we should check process_group_backend is "gloo" or "nccl" in init?
I don't think `init_dist_connection` check that specifically now, if it's invalid backend `torch.distributed.init_process_group` call in will throw exception. If we add to init, it fails faster?

Comment (ananthsub) [reply pytorch_lightning/strategies/ddp.py:186]: No, I do not want to do that because users may pass in custom process group backends like fairring.
